### ISSUE TITLE
Few questions about the training method of OFA

### ISSUE THREAD

**Issue by zhangyxrepo** [LechengKong/OneForAll#23]
Hi, @LechengKong. Thanks for open-sourcing your excellent work. After reading your paper, I still have some questions. And I am here to ask for your kind help.
1.  During the joint training, how do you select the shared parameters of different tasks? And how do you customize the loss function of OFA training on graphs from diverse domains?
2.  Have you ever compared the generalization ability of OFA under different training set mentioned in your paper? Intuitively, it should perform better if train and evaluate on the same dataset, but I am still curious about the generalization ability comparison if it exists.

Thank you again, and looking forward to your reply.

**Comment by LechengKong** [comment]
Hi @zhangyxrepo,

Thank you for liking our work. To answer your questions:

> During the joint training, how do you select the shared parameters of different tasks?

I do not fully understand the question, what do you mean by shared parameters? In OFA, all parameters are shared, meaning that a subgraph from the citation network and a molecular graph will go through exactly the same model and parameter to get their classification results.

> And how do you customize the loss function of OFA training on graphs from diverse domains?

We do not customize loss functions for different tasks. All classification tasks in OFA are carried out in a binary classification on the class node (you can refer to the paper to find the definition of class nodes). To be more concrete, suppose we have two samples in the same batch A and B from different domains with different numbers of classes 2 and 4. In the prompted graph of A, we will have 2 class nodes and 4 for B. Note that despite the prompting process, they are just graphs and we can apply the same GNN on them. After applying the GNN, we get the representations of the class node (2 in the graph of A and 4 in the graph of B), and we use the same MLP to map these representations to a scalar value denoting the likelihood of nodes belonging to the class. Eventually, the class node with the highest scalar value is selected as the class of the node. So, for A we select from 2 scalar values, and for B we select from 4 scalar values.

> Have you ever compared the generalization ability of OFA under different training set mentioned in your paper?

In terms of generalization ability, you can refer to the zero-shot and few-shot experiment, when the training and evaluation datasets are different. OFA is achieving reasonably well performance on these settings.

Hope this helps. If these do not answer your questions, please let me know, happy to elaborate more.

**Comment by zhangyxrepo** [comment]
Hi @LechengKong , thanks for your prompt and clear reply. It helps a lot, but I still have few more questions.

> I do not fully understand the question, what do you mean by shared parameters? In OFA, all parameters are shared, meaning that a subgraph from the citation network and a molecular graph will go through exactly the same model and parameter to get their classification results.

I want to apologize first for the inconvenience caused by my vague question. And your answer is corresponding correctly to what I want to ask. But I want to clarify further, do you mean that in the same training epoch with the same loss function and learned model parameters, the two graphs from domain A and domain B respectively are trained at the same time? Intuitively, this approach introduces a wider data distribution during the training phase, which helps improve the generalization ability of the model, but is there any potential performance degradation reported during your training/evaluating?



> In terms of generalization ability, you can refer to the zero-shot and few-shot experiment, when the training and evaluation datasets are different. OFA is achieving reasonably well performance on these settings.

OFA's generalization ability in this regard is indeed impressive. But I want to ask whether your compared "train on graphs from domain A - zero/few shot test on graphs from domain A" and "train on graphs from domain A and B- zero/few shot test on graphs from domain A or B"? I am curious whether jointly training on cross-domain graphs would impact the performance of OFA or not. 

Thanks again, looking forward to hear from you.

**Comment by LechengKong** [comment]
Ah, that's a good question. Short answer is yes, we train samples from different domains together and it works. To understand this, an analogy is that an LLM is trained on the various domains (math and CommonsenseQA are trained in the same epoch), as you have a large enough embedding space/dimension, the models can learn to separate unrelated domains to eliminate negative transfer while enhancing proximal domains to induce positive transfer. To the second part of the question, you are mostly correct, and there is both performance degradation and enhancement. You can compare the results in Table 2 and Table 3. Row ofa-ind-st is trained only on the training set of the evaluated data, row ofa-st is trained on all training set of tables 2 and 3, if ofa-ind-st > ofa-st, there is degradation, and enhancement otherwise. Most of the time, we see performance enhancement, and we attribute this to two reasons: 1, the input is embedded by high-performance LM, which already well-separates unrelated data and clusters related data. 2, data in the same proximity helps each other (pcba helps hiv, arxiv helps PubMed and cora, etc). To reduce degradation, the most straight forward solution is just using a better LLM, minimizing the chance that downstream GNN confounds two unrelated domains.

The second part is also a good point. If I understand your setting correctly, in the main paper, we follow "train on graphs from domain A and B- zero/few shot test on graphs from domain A or B" setting. The model is jointly trained on FB, ARXIV, chemble. We did "train on graphs from domain A - zero/few shot test on graphs from domain A" setting in table 16 (appendix). We observe similar pattern, joint training usually but not always help.

**Comment by Tianyu9748** [comment]
@LechengKong 
Thank you for open-sourcing this great work. A quick question: LLM is frozen during training or updated?

Best
Tiany

**Comment by LechengKong** [comment]
Hi @Tianyu9748 , the LLM is frozen. In fact, we store the obtained training set node embedding for fast training.

**Comment by Tianyu9748** [comment]
@LechengKong 
Thank you for the response.
